File: Tree/BinaryTree.js
class BinaryTree {
    constructor() {
        this.idx = 0;
    }

    static Node = class {
        constructor(data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }

    buildTree(nodes) {
        if (this.idx >= nodes.length || nodes[this.idx] === -1) return null;
        let root = new BinaryTree.Node(nodes[this.idx]);
        this.idx++;
        root.left = this.buildTree(nodes);
        this.idx++;
        root.right = this.buildTree(nodes);
        return root;
    }

    static preorder(root) {
        if (root === null) return;
        process.stdout.write(root.data + " ");
        BinaryTree.preorder(root.left);
        BinaryTree.preorder(root.right);
    }

    static inorder(root) {
        if (root === null) return;
        BinaryTree.inorder(root.left);
        process.stdout.write(root.data + " ");
        BinaryTree.inorder(root.right);
    }

    static postorder(root) {
        if (root === null) return;
        BinaryTree.postorder(root.left);
        BinaryTree.postorder(root.right);
        process.stdout.write(root.data + " ");
    }

    static levelOrder(root) {
        if (root === null) return;

        let queue = [];
        queue.push(root);
        queue.push(null); // Add null as a marker for the end of the first level

        while (queue.length > 0) {
            let temp = queue.shift();

            if (temp !== null) {
                process.stdout.write(temp.data + " ");
                if (temp.left !== null) queue.push(temp.left);
                if (temp.right !== null) queue.push(temp.right);
            } else {
                // End of current level
                console.log();
                if (queue.length > 0) {
                    queue.push(null); // Add marker for next level
                }
            }
        }
    }
}

// Example usage:
const nodes = [1, 2, 4, 8, -1, -1, 9, -1, -1, 5, 10, -1, -1, 11, -1, -1, 3, 6, -1, -1, 7, -1, -1];
const tree = new BinaryTree();
const root = tree.buildTree(nodes);

console.log("Root data " + root.data);

console.log("Preorder traversal: ");
BinaryTree.preorder(root);
console.log();

console.log("Inorder traversal: ");
BinaryTree.inorder(root);
console.log();

console.log("Postorder traversal: ");
BinaryTree.postorder(root);
console.log();

console.log("Level Order traversal: ");
BinaryTree.levelOrder(root);
console.log();

File: Tree/BinaryTree.c
#include <stdio.h>
#include <stdlib.h>

// Define the structure for the nodes of the binary tree
typedef struct Node {
    int data;
    struct Node* left;
    struct Node* right;
} Node;

// Function to create a new node
Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Function to build the binary tree
Node* buildTree(int nodes[], int* idx, int size) {
    if (*idx >= size || nodes[*idx] == -1) return NULL;
    Node* root = createNode(nodes[*idx]);
    (*idx)++;
    root->left = buildTree(nodes, idx, size);
    (*idx)++;
    root->right = buildTree(nodes, idx, size);
    return root;
}

// Preorder traversal
void preorder(Node* root) {
    if (root == NULL) return;
    printf("%d ", root->data);
    preorder(root->left);
    preorder(root->right);
}

// Inorder traversal
void inorder(Node* root) {
    if (root == NULL) return;
    inorder(root->left);
    printf("%d ", root->data);
    inorder(root->right);
}

// Postorder traversal
void postorder(Node* root) {
    if (root == NULL) return;
    postorder(root->left);
    postorder(root->right);
    printf("%d ", root->data);
}

// Level order traversal
void levelOrder(Node* root) {
    if (root == NULL) return;

    Node* queue[100];
    int front = 0, rear = 0;
    queue[rear++] = root;
    queue[rear++] = NULL; // Add NULL as a marker for the end of the first level

    while (front < rear) {
        Node* temp = queue[front++];

        if (temp != NULL) {
            printf("%d ", temp->data);
            if (temp->left != NULL) queue[rear++] = temp->left;
            if (temp->right != NULL) queue[rear++] = temp->right;
        } else {
            // End of current level
            printf("\n");
            if (front < rear) {
                queue[rear++] = NULL; // Add marker for next level
            }
        }
    }
}

// Main function to test the tree
int main() {
    int nodes[] = {1, 2, 4, 8, -1, -1, 9, -1, -1, 5, 10, -1, -1, 11, -1, -1, 3, 6, -1, -1, 7, -1, -1};
    int size = sizeof(nodes) / sizeof(nodes[0]);
    int idx = 0;

    Node* root = buildTree(nodes, &idx, size);

    printf("Root data %d\n", root->data);

    printf("Preorder traversal: ");
    preorder(root);
    printf("\n");

    printf("Inorder traversal: ");
    inorder(root);
    printf("\n");

    printf("Postorder traversal: ");
    postorder(root);
    printf("\n");

    printf("Level Order traversal:\n");
    levelOrder(root);

    return 0;
}
